pkg/auth/authenticator.go

```diff
@@ -261,3 +271,52 @@ func CheckDomain(domain string, domainAllowedList []string) (bool, error) {
 	}
 	return false, nil
 }
+
+func (a *Authenticator) checkServiceUpstreamAuth(wsvc string, request *Request, extraHeaders *ExtraHeaders) (bool, CerberusReason) {
+	service, ok := (*a.servicesCache)[wsvc]
+	if !ok {
+		return false, CerberusReasonWebserviceNotFound
+	}
+	if service.Spec.UpstreamHttpAuth == (cerberusv1alpha1.UpstreamHttpAuthService{}) {
+		return true, CerberusReasonOK
+	}
+	if service.Spec.UpstreamHttpAuth.ReadTokenFrom == "" {
+		return false, CerberusReasonSourceAuthTokenEmpty
+	}
+	if service.Spec.UpstreamHttpAuth.WriteTokenTo == "" {
+		return false, CerberusReasonTargetAuthTokenEmpty
+	}
+	if !govalidator.IsRequestURL(service.Spec.UpstreamHttpAuth.Address) {
+		return false, CerberusReasonInvalidUpstreamAddress
+	}
+
+	token := request.Request.Header.Get(service.Spec.UpstreamHttpAuth.ReadTokenFrom)
+
+	// TODO: get http method from webservice crd
+	req, err := http.NewRequest("GET", service.Spec.UpstreamHttpAuth.Address, nil)
+	if err != nil {
+		return false, CerberusReasonUpstreamAuthFailed
+	}
+
+	req.Header = http.Header{
+		service.Spec.UpstreamHttpAuth.WriteTokenTo: {token},
+		"Content-Type": {"application/json"},
+	}
+
+	resp, err := a.httpClient.Do(req)
```

Comment:
I think we should support request timeout here (to avoid performance lost)

Reply:
Good point saman.
Now I have added timeout with 200ms default to each webservice defenition.